File: app/src/main/java/fansirsqi/xposed/sesame/ui/MainActivity.java
package fansirsqi.xposed.sesame.ui;

import static fansirsqi.xposed.sesame.data.ViewAppInfo.isApkInDebug;

import android.annotation.SuppressLint;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.core.content.ContextCompat;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import fansirsqi.xposed.sesame.R;
import fansirsqi.xposed.sesame.data.RunType;
import fansirsqi.xposed.sesame.data.UIConfig;
import fansirsqi.xposed.sesame.data.ViewAppInfo;
import fansirsqi.xposed.sesame.entity.FriendWatch;
import fansirsqi.xposed.sesame.entity.UserEntity;
import fansirsqi.xposed.sesame.model.SelectModelFieldFunc;
import fansirsqi.xposed.sesame.util.FansirsqiUtil;
import fansirsqi.xposed.sesame.util.Files;
import fansirsqi.xposed.sesame.util.Log;
import fansirsqi.xposed.sesame.util.Maps.UserMap;
import fansirsqi.xposed.sesame.util.PermissionUtil;
import fansirsqi.xposed.sesame.data.Statistics;
import fansirsqi.xposed.sesame.util.ThreadUtil;
import fansirsqi.xposed.sesame.util.ToastUtil;

public class MainActivity extends BaseActivity {
    private boolean hasPermissions = false;
    private boolean isClick = false;
    private TextView tvStatistics;
    private final Handler viewHandler = new Handler(Looper.getMainLooper());
    private Runnable titleRunner;
    private String[] userNameArray = {"默认"};
    private UserEntity[] userEntityArray = {null};
    private String userId = null;

    @SuppressLint({"UnspecifiedRegisterReceiverFlag", "SetTextI18n"})
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ToastUtil.init(this); // 初始化全局 Context
        hasPermissions = PermissionUtil.checkOrRequestFilePermissions(this);
        if (!hasPermissions) {
            Toast.makeText(this, "未获取文件读写权限", Toast.LENGTH_LONG).show();
            finish(); // 如果权限未获取，终止当前 Activity
            return;
        }
        setContentView(R.layout.activity_main);
        View mainImage = findViewById(R.id.main_image);
        tvStatistics = findViewById(R.id.tv_statistics);
        TextView buildVersion = findViewById(R.id.bulid_version);
        TextView buildTarget = findViewById(R.id.bulid_target);
        TextView oneWord = findViewById(R.id.one_word);
        // 获取并设置一言句子
        ViewAppInfo.checkRunType();
        updateSubTitle(ViewAppInfo.getRunType());
//        viewHandler = new Handler(Looper.getMainLooper());
        titleRunner = () -> updateSubTitle(RunType.DISABLE);
        userId = UserMap.getCurrentUid();
        if (mainImage != null) {
            mainImage.setOnLongClickListener(
                    v -> {
                        // 当视图被长按时执行的操作
                        if (v.getId() == R.id.main_image) {
                            String data = "file://" + Files.getDebugLogFile().getAbsolutePath();
                            Intent it = new Intent(MainActivity.this, HtmlViewerActivity.class);
                            it.putExtra("nextLine", false);
                            it.putExtra("canClear", true);
                            it.setData(Uri.parse(data));
                            startActivity(it);
                            return true; // 表示事件已处理
                        }
                        return false; // 如果不是目标视图，返回false
                    });
        }
        BroadcastReceiver broadcastReceiver =
                new BroadcastReceiver() {
                    @Override
                    public void onReceive(Context context, Intent intent) {
                        String action = intent.getAction();
                        Log.runtime("receive broadcast:" + action + " intent:" + intent);
                        if (action != null) {
                            switch (action) {
                                case "fansirsqi.xposed.sesame.status":
                                    if (RunType.DISABLE == ViewAppInfo.getRunType()) {
                                        updateSubTitle(RunType.PACKAGE);
                                    }
                                    viewHandler.removeCallbacks(titleRunner);
                                    if (isClick) {
                                        // 调用 FansirsqiUtil 获取句子
                                        FansirsqiUtil.getOneWord(
                                                new FansirsqiUtil.OneWordCallback() {
                                                    @Override
                                                    public void onSuccess(String result) {
                                                        runOnUiThread(() -> updateOneWord(result, oneWord)); // 在主线程中更新UI
                                                    }

                                                    @Override
                                                    public void onFailure(String error) {
                                                        runOnUiThread(() -> updateOneWord(error, oneWord)); // 在主线程中更新UI
                                                    }
                                                });
                                        Log.debug("Now User Id: " + userId);
                                        Toast.makeText(context, "芝麻粒状态加载正常👌", Toast.LENGTH_SHORT).show();
                                        ThreadUtil.sleep(200); // 别急，等一会儿再说
                                        isClick = false;
                                    }
                                    break;
                                case "fansirsqi.xposed.sesame.update":
                                    Statistics.load();
                                    tvStatistics.setText(Statistics.getText());
                                    break;
                            }
                        }
                    }
                };
        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction("fansirsqi.xposed.sesame.status");
        intentFilter.addAction("fansirsqi.xposed.sesame.update");
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            registerReceiver(broadcastReceiver, intentFilter, Context.RECEIVER_EXPORTED);
        } else {
            registerReceiver(broadcastReceiver, intentFilter);
        }
        Statistics.load();
        tvStatistics.setText(Statistics.getText());
        // 调用 FansirsqiUtil 获取句子
        FansirsqiUtil.getOneWord(
                new FansirsqiUtil.OneWordCallback() {
                    @Override
                    public void onSuccess(String result) {
                        runOnUiThread(() -> oneWord.setText(result)); // 在主线程中更新UI
                    }

                    @Override
                    public void onFailure(String error) {
                        runOnUiThread(() -> oneWord.setText(error)); // 在主线程中更新UI
                    }
                });
        buildVersion.setText("Build Version: " + ViewAppInfo.getAppVersion()); // 版本信息
        buildTarget.setText("Build Target: " + ViewAppInfo.getAppBuildTarget()); // 编译日期信息
        StringDialog.showAlertDialog(this, "提示", getString(R.string.start_message), "我知道了");
    }

    private void updateOneWord(String str, TextView oneWord) {
        oneWord.setText(str);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (hasPermissions) {
            if (RunType.DISABLE == ViewAppInfo.getRunType()) {
                viewHandler.postDelayed(titleRunner, 3000);
                try {
                    sendBroadcast(new Intent("com.eg.android.AlipayGphone.sesame.status"));
                } catch (Throwable th) {
                    Log.runtime("view sendBroadcast status err:");
                    Log.printStackTrace(th);
                }
            }
            try {//打开设置前需要确认设置了哪个UI
                UIConfig.load();
            } catch (Exception e) {
                Log.printStackTrace(e);
            }
            try {
                List<String> userNameList = new ArrayList<>();
                List<UserEntity> userEntityList = new ArrayList<>();
                java.io.File[] configFiles = Files.CONFIG_DIR.listFiles();
                if (configFiles != null) {
                    for (java.io.File configDir : configFiles) {
                        if (configDir.isDirectory()) {
                            String userId = configDir.getName();
                            UserMap.loadSelf(userId);
                            UserEntity userEntity = UserMap.get(userId);
                            String userName;
                            if (userEntity == null) {
                                userName = userId;
                            } else {
                                userName = userEntity.getShowName() + ": " + userEntity.getAccount();
                            }
                            userNameList.add(userName);
                            userEntityList.add(userEntity);
                        }
                    }
                }
                userNameList.add(0, "默认");
                userEntityList.add(0, null);
                userNameArray = userNameList.toArray(new String[0]);
                userEntityArray = userEntityList.toArray(new UserEntity[0]);
            } catch (Exception e) {
                userNameArray = new String[]{"默认"};
                userEntityArray = new UserEntity[]{null};
                Log.printStackTrace(e);
            }
            try {
                Statistics.load();
                Statistics.updateDay();
                tvStatistics.setText(Statistics.getText());
            } catch (Exception e) {
                Log.printStackTrace(e);
            }
        }
    }

    @SuppressLint("NonConstantResourceId")
    public void onClick(View v) {
        if (v.getId() == R.id.main_image) {
            try {
                sendBroadcast(new Intent("com.eg.android.AlipayGphone.sesame.status"));
                isClick = true;
            } catch (Throwable th) {
                Log.runtime("view sendBroadcast status err:");
                Log.printStackTrace(th);
            }
            return;
        }
        String data = "file://";
        int id = v.getId();
        if (id == R.id.btn_forest_log) {
            data += Files.getForestLogFile().getAbsolutePath();
        } else if (id == R.id.btn_farm_log) {
            data += Files.getFarmLogFile().getAbsolutePath();
        } else if (id == R.id.btn_other_log) {
            data += Files.getOtherLogFile().getAbsolutePath();
        } else if (id == R.id.btn_github) {
            data = "https://github.com/Fansirsqi/Sesame-TK";
        } else if (id == R.id.btn_settings) {
            selectSettingUid();
            return;
        } else if (id == R.id.btn_friend_watch) {
            selectFriendWatchUid();
//            ListDialog.show(this, getString(R.string.friend_watch), FriendWatch.getList(userId), SelectModelFieldFunc.newMapInstance(), false, ListDialog.ListType.SHOW);

            return;
        }
        Intent it = new Intent(this, HtmlViewerActivity.class);
        it.setData(Uri.parse(data));
        startActivity(it);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        PackageManager packageManager = getPackageManager();
        String aliasName = getClass().getCanonicalName() + "Alias";
        try {
            int componentEnabledSetting = packageManager.getComponentEnabledSetting(new ComponentName(this, aliasName));
            MenuItem checkable = menu.add(0, 1, 1, R.string.hide_the_application_icon).setCheckable(true);
            checkable.setChecked(componentEnabledSetting > PackageManager.COMPONENT_ENABLED_STATE_ENABLED);
            menu.add(0, 2, 2, R.string.view_error_log_file);
            menu.add(0, 3, 3, R.string.view_all_log_file);
            menu.add(0, 4, 4, R.string.view_runtim_log_file);
            menu.add(0, 5, 5, R.string.export_the_statistic_file);
            menu.add(0, 6, 6, R.string.import_the_statistic_file);
            menu.add(0, 7, 7, R.string.view_capture);
            menu.add(0, 8, 8, R.string.extend);
            menu.add(0, 9, 9, R.string.settings);
            if (isApkInDebug()) {
                menu.add(0, 10, 10, "Demo Setting UI");
            }
        } catch (Exception e) {
            Log.printStackTrace(e);
            ToastUtil.makeText(this, "菜单创建失败，请重试", Toast.LENGTH_SHORT).show();
        }
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case 1:
                boolean shouldHideIcon = !item.isChecked(); // 是否应隐藏图标
                item.setChecked(shouldHideIcon);
                PackageManager packageManager = getPackageManager();
                String aliasName = getClass().getCanonicalName() + "Alias";
                int newState = shouldHideIcon ? PackageManager.COMPONENT_ENABLED_STATE_DISABLED : PackageManager.COMPONENT_ENABLED_STATE_ENABLED;
                packageManager.setComponentEnabledSetting(new ComponentName(this, aliasName), newState, PackageManager.DONT_KILL_APP);
                break;
            case 2:
                String errorData = "file://";
                errorData += Files.getErrorLogFile().getAbsolutePath();
                Intent errorIt = new Intent(this, HtmlViewerActivity.class);
                errorIt.putExtra("nextLine", false);
                errorIt.putExtra("canClear", true);
                errorIt.setData(Uri.parse(errorData));
                startActivity(errorIt);
                break;
            case 3:
                String RecordData = "file://";
                RecordData += Files.getRecordLogFile().getAbsolutePath();
                Intent otherIt = new Intent(this, HtmlViewerActivity.class);
                otherIt.putExtra("nextLine", false);
                otherIt.putExtra("canClear", true);
                otherIt.setData(Uri.parse(RecordData));
                startActivity(otherIt);
                break;
            case 4:
                String runtimeData = "file://";
                runtimeData += Files.getRuntimeLogFile().getAbsolutePath();
                Intent allIt = new Intent(this, HtmlViewerActivity.class);
                allIt.putExtra("nextLine", false);
                allIt.putExtra("canClear", true);
                allIt.setData(Uri.parse(runtimeData));
                startActivity(allIt);
                break;
            case 5:
                java.io.File statisticsFile = Files.exportFile(Files.getStatisticsFile());
                if (statisticsFile != null) {
                    ToastUtil.makeText(this, "文件已导出到: " + statisticsFile.getPath(), Toast.LENGTH_SHORT).show();
                }
                break;
            case 6:
                if (Files.copyTo(Files.getExportedStatisticsFile(), Files.getStatisticsFile())) {
                    tvStatistics.setText(Statistics.getText());
                    ToastUtil.makeText(this, "导入成功！", Toast.LENGTH_SHORT).show();
                }
                break;
            case 7:
                String captureData = "file://";
                captureData += Files.getCaptureLogFile().getAbsolutePath();
                Intent captureIt = new Intent(this, HtmlViewerActivity.class);
                captureIt.putExtra("nextLine", false);
                captureIt.putExtra("canClear", true);
                captureIt.setData(Uri.parse(captureData));
                startActivity(captureIt);
                break;
            case 8:
                // 扩展功能
                startActivity(new Intent(this, ExtendActivity.class));
                break;
            case 9:
                selectSettingUid();
                break;
            case 10:
                Intent it = new Intent(this, DemoSettingActivity.class);
                it.putExtra("userName", userNameArray[0]);
                startActivity(it);
        }
        return super.onOptionsItemSelected(item);
    }

    public void selectSettingUid() {
        final CountDownLatch latch = new CountDownLatch(1);
        AlertDialog dialog = StringDialog.showSelectionDialog(
                this,
                "请选择配置",
                userNameArray,
                (dialog1, which) -> {
                    goSettingActivity(which);
                    dialog1.dismiss();
                    latch.countDown();
                },
                "返回",
                dialog1 -> {
                    dialog1.dismiss();
                    latch.countDown();
                });

        int length = userNameArray.length;
        if (length > 0 && length < 3) {
            // 定义超时时间（单位：毫秒）
            final long timeoutMillis = 800;
            new Thread(() -> {
                try {
                    if (!latch.await(timeoutMillis, TimeUnit.MILLISECONDS)) {
                        runOnUiThread(() -> {
                            if (dialog.isShowing()) {
                                goSettingActivity(length - 1);
                                dialog.dismiss();
                            }
                        });
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }).start();
        }
    }

    public void selectFriendWatchUid() {
        final CountDownLatch latch = new CountDownLatch(1);
        AlertDialog dialog = StringDialog.showSelectionDialog(
                this,
                "请选择有效账户",
                userNameArray,
                (dialog1, which) -> {
                    goFrinedWatch(which);
                    dialog1.dismiss();
                    latch.countDown();
                },
                "返回",
                dialog1 -> {
                    dialog1.dismiss();
                    latch.countDown();
                });

        int length = userNameArray.length;
        if (length > 0 && length < 3) {
            // 定义超时时间（单位：毫秒）
            final long timeoutMillis = 800;
            new Thread(() -> {
                try {
                    if (!latch.await(timeoutMillis, TimeUnit.MILLISECONDS)) {
                        runOnUiThread(() -> {
                            if (dialog.isShowing()) {
                                goFrinedWatch(length - 1);
                                dialog.dismiss();
                            }
                        });
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }).start();
        }
    }

    private void goFrinedWatch(int index) {
        UserEntity userEntity = userEntityArray[index];
        if (userEntity != null) {
            ListDialog.show(this, getString(R.string.friend_watch), FriendWatch.getList(userEntity.getUserId()), SelectModelFieldFunc.newMapInstance(), false, ListDialog.ListType.SHOW);
        } else {
            ToastUtil.makeText(this, "请先选择有效用户", Toast.LENGTH_SHORT).show();
        }
    }


    /**
     * 启动设置活动，根据用户选择的配置项启动不同的设置界面。
     *
     * @param index 选择的用户索引，用于获取用户信息。
     */
    private void goSettingActivity(int index) {
        UserEntity userEntity = userEntityArray[index];
        Class<?> targetActivity = UIConfig.INSTANCE.getTargetActivityClass();
        Intent intent = new Intent(this, targetActivity);
        if (userEntity != null) {
            intent.putExtra("userId", userEntity.getUserId());
            intent.putExtra("userName", userEntity.getShowName());
        } else {
            intent.putExtra("userName", userNameArray[index]);
        }
        startActivity(intent);
    }

    private void updateSubTitle(RunType runType) {
        setBaseTitle(ViewAppInfo.getAppTitle() + "[" + runType.getName() + "]");
        switch (runType) {
            case DISABLE:
                setBaseTitleTextColor(ContextCompat.getColor(this, R.color.textColorRed));
                break;
            case MODEL:
                setBaseTitleTextColor(ContextCompat.getColor(this, R.color.textColorPrimary));
                break;
            case PACKAGE:
                setBaseTitleTextColor(ContextCompat.getColor(this, R.color.textColorPrimary));
                break;
        }
    }
}

File: app/src/main/java/fansirsqi/xposed/sesame/ui/MainActivity.kt
package fansirsqi.xposed.sesame.ui

import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.MaterialTheme.colorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.ComposeView
import androidx.core.content.ContextCompat
import androidx.core.net.toUri
import androidx.core.util.Consumer
import androidx.lifecycle.lifecycleScope
import fansirsqi.xposed.sesame.R
import fansirsqi.xposed.sesame.data.General
import fansirsqi.xposed.sesame.data.RunType
import fansirsqi.xposed.sesame.data.UIConfig
import fansirsqi.xposed.sesame.data.ViewAppInfo
import fansirsqi.xposed.sesame.entity.FriendWatch
import fansirsqi.xposed.sesame.entity.UserEntity
import fansirsqi.xposed.sesame.model.SelectModelFieldFunc
import fansirsqi.xposed.sesame.ui.widget.ListDialog
import fansirsqi.xposed.sesame.util.AssetUtil
import fansirsqi.xposed.sesame.util.Detector
import fansirsqi.xposed.sesame.util.DeviceInfoUtil
import fansirsqi.xposed.sesame.util.DeviceInfoUtil.DeviceInfoCard
import fansirsqi.xposed.sesame.util.FansirsqiUtil
import fansirsqi.xposed.sesame.util.Files
import fansirsqi.xposed.sesame.util.Log
import fansirsqi.xposed.sesame.util.maps.UserMap
import fansirsqi.xposed.sesame.util.PermissionUtil
import fansirsqi.xposed.sesame.util.ToastUtil
import kotlinx.coroutines.launch
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

//   欢迎自己打包 欢迎大佬pr
//   项目开源且公益  维护都是自愿
//   但是如果打包改个名拿去卖钱忽悠小白
//   那我只能说你妈死了 就当开源项目给你妈烧纸钱了
class MainActivity : BaseActivity() {
    private val TAG = "MainActivity"
    private var hasPermissions = false
    private var userNameArray = arrayOf("默认")
    private var userEntityArray = arrayOf<UserEntity?>(null)
    private lateinit var oneWord: TextView

    @SuppressLint("SetTextI18n", "UnsafeDynamicallyLoadedCode")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ToastUtil.init(this) // 初始化全局 Context
        hasPermissions = PermissionUtil.checkOrRequestFilePermissions(this)
        if (!hasPermissions) {
            Toast.makeText(this, "未获取文件读写权限", Toast.LENGTH_LONG).show()
            finish() // 如果权限未获取，终止当前 Activity
            return
        }
        setContentView(R.layout.activity_main)
        val mainImage = findViewById<View>(R.id.main_image)
        val buildVersion = findViewById<TextView>(R.id.bulid_version)
        val buildTarget = findViewById<TextView>(R.id.bulid_target)
        oneWord = findViewById(R.id.one_word)
        val deviceInfo: ComposeView = findViewById(R.id.device_info)
        deviceInfo.setContent {
            val customColorScheme = lightColorScheme(
                primary = Color(0xFF3F51B5),
                onPrimary = Color.White,
                background = Color(0xFFF5F5F5),
                onBackground = Color.Black
            )
            MaterialTheme(colorScheme = customColorScheme) {
                DeviceInfoCard(DeviceInfoUtil.getDeviceInfo(this@MainActivity))
            }
        }


        // 获取并设置一言句子
        try {
            if (!AssetUtil.copySoFileToStorage(this, AssetUtil.checkerDestFile)) {
                Log.error(TAG, "checker file copy failed")
            }
            if (!AssetUtil.copySoFileToStorage(this, AssetUtil.dexkitDestFile)) {
                Log.error(TAG, "dexkit file copy failed")
            }
            Detector.loadLibrary("checker")
            Detector.initDetector(this)
        } catch (e: Exception) {
            Log.error(TAG, "load libSesame err:" + e.message)
        }

        mainImage?.setOnLongClickListener { v: View ->
            // 当视图被长按时执行的操作
            if (v.id == R.id.main_image) {
                val data = "file://" + Files.getDebugLogFile().absolutePath
                val it = Intent(this@MainActivity, HtmlViewerActivity::class.java)
                it.putExtra("nextLine", false)
                it.putExtra("canClear", true)
                it.data = data.toUri()
                startActivity(it)
                return@setOnLongClickListener true // 表示事件已处理
            }
            false // 如果不是目标视图，返回false
        }
        buildVersion.text = "Build Version: " + ViewAppInfo.appVersion // 版本信息
        buildTarget.text = "Build Target: " + ViewAppInfo.appBuildTarget // 编译日期信息
        lifecycleScope.launch {
            val result = FansirsqiUtil.getOneWord()
            oneWord.text = result
        }
    }

    override fun onResume() {
        super.onResume()
        if (hasPermissions) {
            try { //打开设置前需要确认设置了哪个UI
                UIConfig.load()
            } catch (e: Exception) {
                Log.printStackTrace(e)
            }
            try {
                val userNameList: MutableList<String> = ArrayList()
                val userEntityList: MutableList<UserEntity?> = ArrayList()
                val configFiles = Files.CONFIG_DIR.listFiles()
                if (configFiles != null) {
                    for (configDir in configFiles) {
                        if (configDir.isDirectory) {
                            val userId = configDir.name
                            UserMap.loadSelf(userId)
                            val userEntity = UserMap.get(userId)
                            val userName = if (userEntity == null) {
                                userId
                            } else {
                                userEntity.showName + ": " + userEntity.account
                            }
                            userNameList.add(userName)
                            userEntityList.add(userEntity)
                        }
                    }
                }
                userNameList.add(0, "默认")
                userEntityList.add(0, null)
                userNameArray = userNameList.toTypedArray<String>()
                userEntityArray = userEntityList.toTypedArray<UserEntity?>()
            } catch (e: Exception) {
                userNameArray = arrayOf("默认")
                userEntityArray = arrayOf(null)
                Log.printStackTrace(e)
            }
        }
        updateSubTitle(RunType.LOADED.nickName)
    }

    fun onClick(v: View) {
        if (v.id == R.id.main_image) {
            updateSubTitle(RunType.LOADED.nickName)
            ToastUtil.showToastWithDelay(this, "再点就要去了.~a.e", 800)
            return
        }
        var data = "file://"
        val id = v.id
        when (id) {
            R.id.btn_forest_log -> {
                data += Files.getForestLogFile().absolutePath
            }

            R.id.btn_farm_log -> {
                data += Files.getFarmLogFile().absolutePath
            }

            R.id.btn_other_log -> {
                data += Files.getOtherLogFile().absolutePath
            }

            R.id.btn_github -> {
                data = "https://github.com/Fansirsqi/Sesame-TK"
            }

            R.id.btn_settings -> {
                showSelectionDialog(
                    "📌 请选择配置",
                    userNameArray,
                    { index: Int -> this.goSettingActivity(index) },
                    "😡 老子就不选",
                    {},
                    true
                )
                return
            }

            R.id.btn_friend_watch -> {

                showSelectionDialog(
                    "🤣 请选择有效账户[别选默认]",
                    userNameArray,
                    { index: Int -> this.goFriendWatch(index) },
                    "😡 老子不选了，滚",
                    {},
                    false
                )


                return
            }

            R.id.one_word -> {
                oneWord.text = "😡 正在获取句子，请稍后……"
                lifecycleScope.launch {
                    val result = FansirsqiUtil.getOneWord()
                    oneWord.text = result
                }
                return
            }
        }
        val it = Intent(this, HtmlViewerActivity::class.java)
        it.data = data.toUri()
        startActivity(it)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        try {
            // 使用清单文件中定义的完整别名
            val aliasComponent = ComponentName(this, General.MODULE_PACKAGE_UI_ICON)
            val state = packageManager.getComponentEnabledSetting(aliasComponent)
            // 注意状态判断逻辑修正
            val isEnabled = state != PackageManager.COMPONENT_ENABLED_STATE_DISABLED
            menu.add(0, 1, 1, R.string.hide_the_application_icon)
                .setCheckable(true).isChecked = !isEnabled
            menu.add(0, 2, 2, R.string.view_error_log_file)
            menu.add(0, 3, 3, R.string.view_all_log_file)
            menu.add(0, 4, 4, R.string.view_runtim_log_file)
            menu.add(0, 5, 5, R.string.view_capture)
            menu.add(0, 6, 6, R.string.extend)
            menu.add(0, 7, 7, R.string.settings)
            if (ViewAppInfo.isApkInDebug) {
                menu.add(0, 8, 8, "清除配置")
            }
        } catch (e: Exception) {
            Log.printStackTrace(e)
            ToastUtil.makeText(this, "菜单创建失败，请重试", Toast.LENGTH_SHORT).show()
            return false
        }
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            1 -> {
                val shouldHide = !item.isChecked
                item.isChecked = shouldHide

                val aliasComponent = ComponentName(this, General.MODULE_PACKAGE_UI_ICON)
                val newState = if (shouldHide) {
                    PackageManager.COMPONENT_ENABLED_STATE_DISABLED
                } else {
                    PackageManager.COMPONENT_ENABLED_STATE_ENABLED
                }

                packageManager.setComponentEnabledSetting(
                    aliasComponent,
                    newState,
                    PackageManager.DONT_KILL_APP
                )

                // 提示用户需要重启启动器才能看到效果
                Toast.makeText(this, "设置已保存，可能需要重启桌面才能生效", Toast.LENGTH_SHORT)
                    .show()
                return true
            }

            2 -> {
                var errorData = "file://"
                errorData += Files.getErrorLogFile().absolutePath
                val errorIt = Intent(this, HtmlViewerActivity::class.java)
                errorIt.putExtra("nextLine", false)
                errorIt.putExtra("canClear", true)
                errorIt.data = errorData.toUri()
                startActivity(errorIt)
            }

            3 -> {
                var recordData = "file://"
                recordData += Files.getRecordLogFile().absolutePath
                val otherIt = Intent(this, HtmlViewerActivity::class.java)
                otherIt.putExtra("nextLine", false)
                otherIt.putExtra("canClear", true)
                otherIt.data = recordData.toUri()
                startActivity(otherIt)
            }

            4 -> {
                var runtimeData = "file://"
                runtimeData += Files.getRuntimeLogFile().absolutePath
                val allIt = Intent(this, HtmlViewerActivity::class.java)
                allIt.putExtra("nextLine", false)
                allIt.putExtra("canClear", true)
                allIt.data = runtimeData.toUri()
                startActivity(allIt)
            }

            5 -> {
                var captureData = "file://"
                captureData += Files.getCaptureLogFile().absolutePath
                val captureIt = Intent(this, HtmlViewerActivity::class.java)
                captureIt.putExtra("nextLine", false)
                captureIt.putExtra("canClear", true)
                captureIt.data = captureData.toUri()
                startActivity(captureIt)
            }

            6 ->                 // 扩展功能
                startActivity(Intent(this, ExtendActivity::class.java))

            7 -> selectSettingUid()
            8 -> AlertDialog.Builder(this)
                .setTitle("⚠️ 警告")
                .setMessage("🤔 确认清除所有模块配置？")
                .setPositiveButton(R.string.ok) { _: DialogInterface?, _: Int ->
                    if (Files.delFile(Files.CONFIG_DIR)) {
                        Toast.makeText(this, "🙂 清空配置成功", Toast.LENGTH_SHORT).show()
                    } else {
                        Toast.makeText(this, "😭 清空配置失败", Toast.LENGTH_SHORT).show()
                    }
                }
                .setNegativeButton(R.string.cancel) { dialog: DialogInterface, _: Int -> dialog.dismiss() }
                .create()
                .show()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun selectSettingUid() {
        val latch = CountDownLatch(1)
        val dialog = StringDialog.showSelectionDialog(
            this,
            "📌 请选择配置",
            userNameArray,
            { dialog1: DialogInterface, which: Int ->
                goSettingActivity(which)
                dialog1.dismiss()
                latch.countDown()
            },
            "返回",
            { dialog1: DialogInterface ->
                dialog1.dismiss()
                latch.countDown()
            })

        val length = userNameArray.size
        if (length in 1..2) {
            // 定义超时时间（单位：毫秒）
            val timeoutMillis: Long = 800
            Thread {
                try {
                    if (!latch.await(timeoutMillis, TimeUnit.MILLISECONDS)) {
                        runOnUiThread {
                            if (dialog.isShowing) {
                                goSettingActivity(length - 1)
                                dialog.dismiss()
                            }
                        }
                    }
                } catch (_: InterruptedException) {
                    Thread.currentThread().interrupt()
                }
            }.start()
        }
    }

    private fun showSelectionDialog(
        title: String?, options: Array<String>,
        onItemSelected: Consumer<Int>,
        negativeButtonText: String?,
        onNegativeButtonClick: Runnable,
        showDefaultOption: Boolean
    ) {
        val latch = CountDownLatch(1)
        val dialog = StringDialog.showSelectionDialog(
            this,
            title,
            options,
            { dialog1: DialogInterface, which: Int ->
                onItemSelected.accept(which)
                dialog1.dismiss()
                latch.countDown()
            },
            negativeButtonText,
            { dialog1: DialogInterface ->
                onNegativeButtonClick.run()
                dialog1.dismiss()
                latch.countDown()
            })

        val length = options.size
        if (showDefaultOption && length > 0 && length < 3) {
            // 定义超时时间（单位：毫秒）
            val timeoutMillis: Long = 800
            Thread {
                try {
                    if (!latch.await(timeoutMillis, TimeUnit.MILLISECONDS)) {
                        runOnUiThread {
                            if (dialog.isShowing) {
                                onItemSelected.accept(length - 1)
                                dialog.dismiss()
                            }
                        }
                    }
                } catch (_: InterruptedException) {
                    Thread.currentThread().interrupt()
                }
            }.start()
        }
    }


    private fun goFriendWatch(index: Int) {
        val userEntity = userEntityArray[index]
        if (userEntity != null) {
            ListDialog.show(
                this,
                getString(R.string.friend_watch),
                FriendWatch.getList(userEntity.userId),
                SelectModelFieldFunc.newMapInstance(),
                false,
                ListDialog.ListType.SHOW
            )
        } else {
            ToastUtil.makeText(this, "😡 别他妈选默认！！！！！！！！", Toast.LENGTH_LONG).show()
        }
    }

    private fun goSettingActivity(index: Int) {
        if (Detector.loadLibrary("checker")) {
            val userEntity = userEntityArray[index]
            val targetActivity = UIConfig.INSTANCE.targetActivityClass
            val intent = Intent(this, targetActivity)
            if (userEntity != null) {
                intent.putExtra("userId", userEntity.userId)
                intent.putExtra("userName", userEntity.showName)
            } else {
                intent.putExtra("userName", userNameArray[index])
            }
            startActivity(intent)
        } else {
            Detector.tips(this, "缺少必要依赖！")
        }
    }

    fun updateSubTitle(runType: String) {
        Log.runtime(TAG, "updateSubTitle$runType")
        baseTitle = ViewAppInfo.appTitle + "[" + runType + "]"
        when (runType) {
            RunType.DISABLE.nickName -> setBaseTitleTextColor(
                ContextCompat.getColor(
                    this,
                    R.color.not_active_text
                )
            )

            RunType.ACTIVE.nickName -> setBaseTitleTextColor(
                ContextCompat.getColor(
                    this,
                    R.color.active_text
                )
            )

            RunType.LOADED.nickName -> setBaseTitleTextColor(
                ContextCompat.getColor(
                    this,
                    R.color.textColorPrimary
                )
            )
        }
    }
}
